src/interface/validation.rs

```diff
@@ -0,0 +1,252 @@
+// This file is part of Astarte.
+// Copyright 2023 SECO Mind Srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Validate an interface and the version change between two interfaces.
+
+use std::{cmp::Ordering, fmt::Display};
+
+use super::{
+    error::{Error, ValidationError, VersionChangeError},
+    Interface,
+};
+
+/// A change in version of an interface.
+#[derive(Debug, Clone, Copy)]
+pub struct VersionChange {
+    pub next_major: i32,
+    pub next_minor: i32,
+    pub prev_major: i32,
+    pub prev_minor: i32,
+}
+
+impl VersionChange {
+    pub fn new(next: &Interface, prev: &Interface) -> Self {
+        Self {
+            next_major: next.get_version_minor(),
+            next_minor: next.get_version_minor(),
+            prev_major: prev.get_version_major(),
+            prev_minor: prev.get_version_minor(),
+        }
+    }
+
+    pub fn validate(&self) -> Result<(), VersionChangeError> {
+        let majior = self.next_major.cmp(&self.prev_major);
+        let minor = self.next_minor.cmp(&self.prev_minor);
+
+        match (majior, minor) {
+            (Ordering::Less, _) => Err(VersionChangeError::MajorDecresed(*self)),
+            (Ordering::Equal, Ordering::Less) => Err(VersionChangeError::MinorDecresed(*self)),
+            (Ordering::Equal, Ordering::Equal) => Err(VersionChangeError::SameVersion(*self)),
+            _ => Ok(()),
+        }
+    }
+}
+
+impl Display for VersionChange {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "next: {}.{} prev: {}.{}",
+            self.next_major, self.next_minor, self.prev_major, self.prev_minor
+        )
+    }
+}
+
+impl Interface {
```

Comment:
I wanted to encapsulate all the validation logic with the unit test in a single file, thinking this would make the interface/mod.rs clearer/simpler. 

I would prefer to keep them this way, but if you think it would be cleared I can:
- move the only the `impl` in the mod.rs
- or move the `impl` + unit test in the mod.rs, but then this file would be pretty empty

What do you think?